classes/ocis_manager.php

```diff
@@ -0,0 +1,496 @@
+            return [];
+        } else {
+            $proxyhost = $CFG->proxyhost;
+            if (!empty($CFG->proxyport)) {
+                $proxyhost = "{$CFG->proxyhost}:{$CFG->proxyport}";
+            }
+
+            $proxyauth = "";
+            if (!empty($CFG->proxyuser) && !empty($CFG->proxypassword)) {
+                $proxyauth = "{$CFG->proxyuser}:{$CFG->proxypassword}@";
+            }
+
+            $protocol = "http://";
+            if (!empty($CFG->proxytype) && $CFG->proxytype === 'SOCKS5') {
+                $protocol = "socks5://";
+            }
+
+            $proxystring = "{$protocol}{$proxyauth}{$proxyhost}";
+            $noproxy = [];
+
+            if (!empty($CFG->proxybypass)) {
+                $noproxy = array_map(function (string $hostname): string {
+                    return trim($hostname);
+                }, explode(',', $CFG->proxybypass));
+            }
+
+            return [
+                'http' => $proxystring,
+                'https' => $proxystring,
+                'no' => $noproxy,
+            ];
+        }
+    }
+
+    private function get_new_ocis_object(
+        string $accesstoken,
+        array $proxysetting
+    ): Ocis {
+        $webfingerurl = issuer_management::get_webfinger_url($this->oauth2issuer);
+        if ($webfingerurl) {
+            try {
+                return new Ocis(
+                    $webfingerurl,
+                    $accesstoken,
+                    [
+                        'webfinger' => true,
+                        'proxy' => $proxysetting,
+                    ]
+                );
+            } catch (InternalServerErrorException $e) {
+                throw new moodle_exception(
+                    'internal_server_error',
+                    'repository_ocis',
+                    '',
+                    null,
+                    $e->getTraceAsString()
+                );
+            } catch (\Exception $e) {
+                throw new moodle_exception(
+                    'webfinger_error',
+                    'repository_ocis',
+                    '',
+                    null,
+                    $e->getMessage()
+                );
+            }
+        } else {
+            $baseurl = $this->oauth2issuer->get('baseurl');
+            return new Ocis($baseurl, $accesstoken, ['proxy' => $proxysetting]);
+        }
+    }
+
+    public function set_driveid_and_path(string $driveidandpath) {
+        $this->driveidandpath = $driveidandpath;
+    }
+
+    private function is_root() {
+        return ($this->driveidandpath === '' || $this->driveidandpath === '/');
+    }
+
+    public function get_ocis_client(): Ocis {
+        $accesstoken = $this->oauth2client->get_accesstoken();
+        if ($accesstoken === null) {
+            throw new moodle_exception(
+                'unauthorized_error',
+                'repository_ocis'
+            );
+        }
+        if ($this->ocis === null) {
```

Comment:
My IDE tells me that `$this->ocis` is being declared dynamically here.
IMO it needs to be declared as a property in this class.